File: seconds_to_sentence.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Given n seconds, return a 'x hours, y minutes and n seconds' sentence.
#
class IntegerTime
  UNITS = { day: 86_400, hour: 3600, minute: 60, second: 1 }.freeze

  # @param seconds [Integer] number of seconds to convert
  # @return [String]
  #
  def as_a_sentence(seconds)
    return 'Unsupported input.' unless seconds.is_a?(Integer) && seconds >= 0
    return 'No time at all.' if seconds.zero?

    make_sentence(pluralise(chunks(seconds)))
  end

  private

  # @param seconds [Integer] number of seconds to convert
  # @return [Array[Array]] with elements of the form ['day', 10]
  #
  def chunks(seconds)
    UNITS.each_with_object([]) do |(word, value), a|
      units = seconds / value
      seconds = seconds % value
      a.<<([word.to_s, units]) if units.positive?
    end
  end

  # @param arr [Array[Array]] with elements of the form ['day', 10]
  # @return [Array[String]] with elements of the form '10 days'
  #
  def pluralise(arr)
    arr.map do |word, num|
      word += 's' unless num == 1
      "#{num} #{word}"
    end
  end

  # @param arr [Array[String]] with elements of the form '10 days'
  # @return [String] of the form '10 days and 4 seconds.'
  #
  def make_sentence(arr)
    arr.each.with_index(2).map do |chunk, index|
      chunk += ', ' if index < arr.size
      index == arr.size ? chunk + ' and ' : chunk
    end.join + '.'
  end
end

# Run with TEST as the only arg to test. The block of tests is nicked from the
# pytest block in the Python version. I wouldn't do this normally.
#
if ARGV.first == 'TEST'
  def assert(input, expected)
    raise unless IntegerTime.new.as_a_sentence(input) == expected
  end

  assert 0, 'No time at all.'
  assert 59, '59 seconds.'
  assert 60, '1 minute.'
  assert 3602, '1 hour and 2 seconds.'
  assert 86_400, '1 day.'
  assert 100_000, '1 day, 3 hours, 46 minutes and 40 seconds.'
  assert(-1, 'Unsupported input.')
  assert '0', 'Unsupported input.'
  assert 0.4, 'Unsupported input.'
  puts 'all tests passed'
  exit
end

if ARGV.first != ARGV.first.to_i.to_s || ARGV.first.to_i.negative?
  abort 'Please supply a positive integer as the only argument.'
end

puts IntegerTime.new.as_a_sentence(ARGV.first.to_i)

File: seconds_to_sentence.py
#!/usr/bin/env python3

import sys

# Homework for ChrisA's Python classes.
# Print a given positive integer number of seconds as an English sentence.


class IntegerTime:
    Units = {'day': 86400,
             'hour': 3600,
             'minute': 60,
             'second': 1}

    def as_a_sentence(self, timestamp):
        if not isinstance(timestamp, int) or timestamp < 0:
            raise('Unsupported input.')

        if timestamp == 0:
            return 'No time at all.'

        return self.__make_sentence(self.__pluralise(self.__chunks(timestamp)))

    def __chunks(self, seconds):
        ret = []

        for word, value in self.Units.items():
            units = seconds // value
            seconds = seconds % value

            if units > 0:
                ret.append([word, units])

        return ret

    def __pluralise(self, arr):
        ret = []

        for word, num in arr:
            if num != 1:
                word += 's'

            ret.append(f'{num} {word}')

        return ret

    def __make_sentence(self, arr):
        ret_string = ''
        word_index = 1
        number_of_words = len(arr)

        for word in arr:
            word_index += 1
            ret_string += word

            if word_index < number_of_words:
                ret_string += ', '
            elif word_index == number_of_words:
                ret_string += ' and '

        return ret_string + '.'


# Test public interface of IntegerTime class

def test_as_a_sentence():
    assert 'No time at all.' == IntegerTime().as_a_sentence(0)
    assert '59 seconds.' == IntegerTime().as_a_sentence(59)
    assert '1 minute.' == IntegerTime().as_a_sentence(60)
    assert '1 hour and 2 seconds.' == IntegerTime().as_a_sentence(3602)
    assert '1 day.' == IntegerTime().as_a_sentence(86400)
    assert '1 day, 3 hours, 46 minutes and 40 seconds.' == \
        IntegerTime() .as_a_sentence(100000)
    assert 'Unsupported input.' == IntegerTime().as_a_sentence(-1)
    assert 'Unsupported input.' == IntegerTime().as_a_sentence('0')
    assert 'Unsupported input.' == IntegerTime().as_a_sentence(0.4)


try:
    user_input = int(sys.argv[1])
    print(IntegerTime().as_a_sentence(user_input))
except ValueError:
    print('Please supply a positive integer as the only argument.')
    sys.exit(1)
